### DISCUSSION TITLE
Pickle dumping & loading causes model learn_one to forget all history

### DISCUSSION THREAD

**Discussion by mshussein** [online-ml/river#861] [General]
Hello,

I'll just preface this by saying I'm not a data scientist, just a seasoned web developer who's trying to expand horizons.

I've made a rudimentary bank transaction classifier that infers the type of transaction based on the description and amount. Eg shopping, loans, dining out, etc.

I've defined my pipeline as follows:
```
model = Pipeline(
    ('vectorizer', BagOfWords('Description',lowercase=True)),
    ('model', river.multiclass.OneVsOneClassifier(river.linear_model.PAClassifier())),
)
```

I then train it from a CSV and all works quite well!

Something like this:
```
for index, row in transactionsSample.iterrows():
    model = model.learn_one({'Description':row[1], 'Amount':row[2]}, row[0])
```
I need to dump this to file and load it on a container on AWS to serve requests for both training and prediction.

Now, if I ever save & reload it using pickle, and then do a single model.learn_one() on it, ALL other (previously correct) predictions result in the same answer as provided to the single model.learn_one function.

**Save:**
```
with open('model.pkl', 'wb') as f:
    pickle.dump(model, f)
```

**Load:**
```
with open('model.pkl', 'rb') as f:
    model = pickle.load(f)
```
From the documentation (https://riverml.xyz/latest/api/compose/Pipeline/), I see that if you're doing learn_one on a pipeline, it fits the model to this new value, quote: "Fit to a single instance." 

Other learn_one functions (eg https://riverml.xyz/latest/api/multiclass/OneVsOneClassifier/) say: "Update the model with a set of features x and a label y."

My gut feel is that it's not saving/loading correctly, therefore losing all historical learning, making the next learn_one the *only* training it has ever received.

**Questions:**
1. Am I saving & loading the model correctly?
2. Am I training it correctly? Both in the initial CSV training and the subsequent training post-load? (both use the pipeline to learn_one so I can't understand why it's any different)
3. What am I doing wrong?

I appreciate the help. I'm totally lost.

**Comment by mshussein** [comment]
Okay,

Further to the above, I find the issue happens even when I don't save and reload!

If I train a CSV model, run some predictions, then learn_one immediately afterwards (once), all previous predictions when re-done have the answer of my latest learning instead

**Reply by MaxHalford** [reply]
This can happen if your dataset is not shuffled. That is, it might be ordered by class. This makes the learning process to trivial and doesn't in fact result in a model that has learned anything. Could you check if the same problem occurs if your dataset is shuffled?

**Reply by mshussein** [reply]
Thanks Max,

Yes still happens. They're not sorted by class, but I took another sample of my training sample, just to make sure my data preprocessing didn't inadvertently sort it by class, and it still happens.

All in all, it works well, and I achieve about 87% accuracy from training only about 15,000 transactions. I have 28 million to play with should I need.

I fail to understand how its happening.

I stripped out all the fluff, and if we consider this chunk of pseudocode:

```
#read all 28 million transactions
transactions = pd.read_csv(...) 

#take a sample
transactionsSampleFull = transactions.sample(frac=0.001,random_state=200)

# remove the catch all categories for both debits and credits
transactionsSample = pd.DataFrame(transactionsSampleFull[(transactionsSampleFull.Category != "Other debit") & (transactionsSampleFull.Category != "All Other Credits")]) 

#remove characters i dont want
transactionsSample['Description'] = transactionsSample['Description'].str.upper().replace('[\d+,\(,\),_,-]', '', regex=True) 

# get a sample of the sample to shuffle it
transactionsSample = transactionsSample.sample(frac=1, random_state=100)

# define pipeline
model = Pipeline(
    ('vectorizer', BagOfWords('Description',lowercase=True)),
    ('model', river.multiclass.OneVsOneClassifier(river.linear_model.PAClassifier())),
)

# train
for index, row in transactionsSample.iterrows():
    model = model.learn_one({'Description':row['Description'], 'Amount':row['Amount']}, row['Category'])

# do some random predictions here. These will be pretty accurate
print(model.predict_one({'Description': 'DOMINOs', 'Amount': -5.5}))
print(model.predict_one({'Description': 'DOMINOs', 'Amount': -16.5}))
print(model.predict_one({'Description': 'BURGERME', 'Amount': -22.93}))
print(model.predict_one({'Description': 'WOOLWORTHS      ', 'Amount': -39.6}))
print(model.predict_one({'Description': 'C61880 TFR FROM Westpa c Choice', 'Amount': 278.46}))
print(model.predict_one({'Description': 'MR TOYS', 'Amount': -5.99}))

#the moment i run this below, any prediction afterwards is 'Income', before then, predictions are totally fine? 90% accurate before this line
# one would imagine this would be equivalent to having 1 more row in my transactionsSample, but it's not
model = model.learn_one({'Description':'Salary From Work','Amount':589.05}, 'Income')

# these will all be Income 
print(model.predict_one({'Description': 'DOMINOs', 'Amount': -5.5}))
print(model.predict_one({'Description': 'DOMINOs', 'Amount': -16.5}))
print(model.predict_one({'Description': 'BURGERME', 'Amount': -22.93}))
print(model.predict_one({'Description': 'WOOLWORTHS      ', 'Amount': -39.6}))
print(model.predict_one({'Description': 'C61880 TFR FROM Westpa c Choice', 'Amount': 278.46}))
print(model.predict_one({'Description': 'MR TOYS', 'Amount': -5.99}))
```

I'll try and get some anonymized transactions if you want to try it yourself. Let me know.

An aside: when training is it necessary to reassign the return value of learn_one back to the model?

model  = model.learn_one
**vs**
model.learn_one

Kind regards.

**Reply by MaxHalford** [reply]
I don't have time to dig into this right now. Maybe next week. In the meantime, could you try a different base classifier? For example, I suggest `LogisticRegression` instead of `PAClassifier`.

**Reply by mshussein** [reply]
Thanks @MaxHalford,

No problem at all, thanks for your time so far. Much appreciated.

As requested, I changed it to:
````
model = Pipeline(
    ('vectorizer', TFIDF(on='Description',lowercase=True)),
    ('model', river.multiclass.OneVsOneClassifier(river.linear_model.LogisticRegression())),
)
````
(I got an error trying to use BagOfWords, "Expected dict, got Counter", so I switched to TFIDF just to perform your test. It's not as accurate as BagOfWords for the data I have, but I just wanted to use LogisticRegression for the classifier)

Either way, still happens as described: As soon as I learn_one outside of the for loop, all results revert to the last trained answer.

**Reply by mshussein** [reply]
Sorry mate,

**I'm such a muppet.**

I figured out the extra learn_one was a new transaction category. I did make up 2 transactions, but they both coincidentally happened to be categories previously unseen in my (small) sample of about 8000. (Sorry!)

I can confirm when we do a previously known category, we're all good.

That being said, I suppose if we do need to add a new category, we may need to do a lot more training of the new and old categories, lest the newly added category becomes the default answer.

I don't know if that helps at all, or if there's still an issue. I would imagine however that learn_one on a brand new category shouldn't do what it's doing?

Your time and efforts are very much appreciated.

Kind regards.
Muhammed.

**Reply by MaxHalford** [reply]
It's ok, mistakes happen! I'm glad you figured it out.

> That being said, I suppose if we do need to add a new category, we may need to do a lot more training of the new and old categories, lest the newly added category becomes the default answer.

Depends. Retraining can be good. But normally you're learning from a stream of data, so if a new category is predominant in a stream, then there's no need to be good at classifying the old categories.

**Reply by mshussein** [reply]
I guess my use case is a little different than streaming data.

It would be to allow our staff to correct any mistakes in transaction categorization if/when we encounter them.

As long as we don't have to add a brand new category and we are only correcting, I think we'll be okay.

Thanks again for taking the time and I do apologize if I have wasted any!

**Reply by mshussein** [reply]
Also, even further to this in case this helps anyone else:

- Training a new category causes all outputs to be that single category
- Training something that results in a previously seen category, reverts all outputs to what they were before (the correct ones)
